### PR TITLE
Add USPS rate estimation endpoint

### PR DESCRIPTION
**Pull request by mamhoff**
This adds the capability of estimating USPS rates directly from USPS.
There is pretty complicated code, because the USPS rating API has a lot
of functionality and an interesting data model.

When requesting rates for a shipment, USPS will return all possible
rates for each package. This means that we need to match the response's
rates to each of the packages, and then sum the rates for all the
packages to get rates for the entire shipment.

Compared to previous API implementations, we're using USPS' endpoint
with POST requests rather than GET requests because that makes VCR
cassette matching easier and probably also allows longer a longer
request payload.

### CODE REVIEW

File: lib/friendly_shipping/services/usps/parse_rate_response.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+          # @param [Nokogiri::XML::Node] xml The XML document containing packages and rates
+          # @param [Physical::Shipment] shipment The shipment we're trying to get rates for
+          #
+          # @return [Hash<Physical::Package => Array<FriendlyShipping::Rate>>]
+          def rates_from_response_node(xml, shipment)
+            xml.xpath(PACKAGE_NODE_XPATH).each_with_object({}) do |package_node, result|
+              package_id = package_node['ID']
+              corresponding_package = shipment.packages.detect { |p| p.id == package_id }
+              result[corresponding_package] = package_node.xpath(SERVICE_NODE_NAME).map do |service_node|
+                ParseRate.call(service_node, corresponding_package)
```
Comment (pelargir) [inline lib/friendly_shipping/services/usps/parse_rate_response.rb]: How about `ParsePackageRate`? Something to distinguish it from `ParseRateResponse` a bit more. To me, `ParseRateResponse` and `ParseRate` could be confused as the same thing.

File: lib/friendly_shipping/services/usps/serialize_rate_request.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require 'friendly_shipping/services/usps/machinable_package'
+
+module FriendlyShipping
+  module Services
+    class Usps
+      class SerializeRateRequest
```
Comment (pelargir) [inline lib/friendly_shipping/services/usps/serialize_rate_request.rb]: I think we need a spec for this class. One that tests the usual First Class or Priority shipping methods as well as a Regional Rate shipping method, checking that Service=PRIORITY COMMERCIAL and Container=REGIONALRATEBOXA (or whatever is used).

File: spec/friendly_shipping/services/usps/choose_package_rate_spec.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  let(:package_id) { '0' }
+  let(:package) { FactoryBot.build(:physical_package, id: package_id) }
+  let(:xml) { File.open(File.join(gem_root, 'spec', 'fixtures', 'usps', 'usps_rates_api_response.xml')).read }
+  let(:rate_nodes) { Nokogiri::XML(xml).xpath('//Postage') }
+  let(:rates) { rate_nodes.map { |node| FriendlyShipping::Services::Usps::ParseRate.call(node, package) } }
+
+  subject { described_class.call(shipping_method, package, rates) }
+
+  it { is_expected.to be_a(FriendlyShipping::Rate) }
+end
```
Comment (tvdeyen) [inline spec/friendly_shipping/services/usps/choose_package_rate_spec.rb]: I think you should add some spes for the "hold for pickup" and "wrong box type" cases you address in the class.

File: spec/friendly_shipping/services/usps/parse_xml_response_spec.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  subject(:parser) { described_class.call(response, 'SomeTag') }
+
+  it { is_expected.to be_failure }
+
+  it 'has the correct error message' do
+    expect(subject.failure.to_s).to eq('0: Something went wrong')
+  end
+end
```
Comment (tvdeyen) [inline spec/friendly_shipping/services/usps/parse_xml_response_spec.rb]: Please add a spec for a successful response as well.

File: lib/friendly_shipping/services/usps/parse_rate.rb [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+            # Clean up the long string
+            service_name.gsub!(SERVICE_NAME_SUBSTITUTIONS, '')
+
+            # Some USPS services only offer commercial pricing. Unfortunately, USPS then returns a retail rate of 0.
+            # In these cases, return the commercial rate instead of the normal rate.
+            # Some rates are available in both commercial and retail pricing - if we want the commercial pricing here,
+            # we need to specify the commercial_pricing property on the `Physical::Package`.
+            rate_value = if (package.properties[:commercial_pricing] || rate_node.at(RATE_TAG).text.to_d.zero?) && rate_node.at(COMMERCIAL_RATE_TAG)
+                           rate_node.at(COMMERCIAL_RATE_TAG)&.text.to_d
```
[the remark this answers is not in the document]
Comment (mamhoff) [reply lib/friendly_shipping/services/usps/parse_rate.rb]: Hm, actually the safe navigation operator isn't necessary here - we only enter this branch if the `COMMERCIAL_RATE_TAG` is present.

File: spec/friendly_shipping/services/usps/parse_xml_response_spec.rb (hunk 4 above)
Comment (mamhoff) [reply spec/friendly_shipping/services/usps/parse_xml_response_spec.rb]: Done, good call - adding this spec revealed (and fixed) bugs.

File: lib/friendly_shipping/services/usps/shipping_methods.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+      SHIPPING_METHODS = [
+        'First-Class Mail',
+        'First-Class Package Service',
+        'Priority Mail',
+        'Priority Mail Express',
+        'Standard Post',
+        'Retail Ground',
+        'Media Mail',
+        'Library Mail',
+      ].freeze.map do |shipping_method_name|
```
[the remark this answers is not in the document]
Comment (mamhoff) [reply lib/friendly_shipping/services/usps/shipping_methods.rb]: Thx

File: lib/friendly_shipping/services/usps/serialize_rate_request.rb (hunk 2 above)
Comment (mamhoff) [reply lib/friendly_shipping/services/usps/serialize_rate_request.rb]: Yeah! I forgot pulling the spec for this one, thank you for reminding me. Spec is now added. 

You will notice that we don't automatically quote commercial now if a flat rate box is used, but that that has to be decided from the calling code.

File: lib/friendly_shipping/services/usps/parse_rate_response.rb (hunk 1 above)
Comment (mamhoff) [reply lib/friendly_shipping/services/usps/parse_rate_response.rb]: Yeah, I can do that.

**Pull request merged**